### ISSUE TITLE
upgrade to adonis2

### ISSUE THREAD

**Issue by nbokulich** [qiime2/q2-diversity#328]
**Improvement Description**
Upgrade to the vegan function `adonis2` to access additional functionality unavailable in adonis1.

To quote @mestaki :
>The most important benefit would be that adonis2 allows marginal testing of variables using the by="marginal" parameter and the output gives a partial R-squared for each variable in the model.

**Current Behavior**
Currently `adonis` uses adonis1 in vegan. This lacks some functionality of adonis2.

To quote @mestaki :
> the current QIIME 2 version of adonis can only perform a sequential test of terms (equivalent to by="terms" in adonis2(), meaning the order of the variables in the formula is important with the residuals from the first term being passed on to the second term, then residuals from the 2nd to 3rd and so on. I suspect most users are not aware of this and it is easy to miss an important biological signal in your model if a term of interest is unintentionally shoved at the end of the formula.

**Proposed Behavior**
Swap the functions, but keep the defaults to match the current adonis1 functionality (this should be possible with adonis2 surely).

Also expose marginal testing and other options available in `adonis2` (see related open issues about adonis parameters, e.g., #242)

**References**
Raised by @mestaki on the [forum](https://forum.qiime2.org/t/updating-to-adonis2/23147?u=nicholas_bokulich)

**Comment by jwdebelius** [comment]
I would love if this refactor (which I'm happy to help with) output an artifact rather than a visualization. The results could then be passed into the current visualizer, but it would make adonis/adonis2 easier to use in the python interface for multiple testing or for comparing models.

Again, happy to help!

**Comment by ebolyen** [comment]
Hey @jwdebelius, @lizgehret and I have been working on reporting statistical outputs as artifacts (in no small part because of your suggestions at the last workshop). If you'd like to see how we've been doing it, you can find it here: https://github.com/qiime2/q2-fmt , we'd be super interested in any feedback you can give.

We're using [Tabular Data Resources](https://specs.frictionlessdata.io/tabular-data-resource/) as the backing format because it makes the data a bit more standardized for alternate environments and lets us attach metadata to the columns, so we can preserve things like labels and other info (ideally mirroring the convenience of R a little bit).

Maybe these can serve as a template for the Adonis refactor? It may be worth having an offline conversation about what our plans are in the near-term for this stuff. I think it would be really great to have you on-board! (cc @gregcaporaso and @nbokulich)

**Comment by jwdebelius** [comment]
Thanks @ebolyen,

I will check it out! Thank you. And I'm happy to talk offline and contribute, if I can. I'm not too familiar with the tabular data resource.

**Comment by mestaki** [comment]
Thanks @nbokulich for starting this and @jwdebelius for offering to help it out!
I'm happy to help anyway I can as well but will be nowhere as effective as Justine lol.

> Proposed Behavior
Swap the functions, but keep the defaults to match the current adonis1 functionality (this should be possible with adonis2 surely).

Yes, totally, there are a few options here. The updated vegan package comes with both adonis() and adonis2(). Option 1: keep default as adonis() to avoid any backward compatibility, and add adonis2 as a separate plugin. Option2: use adonis() as default and switch to adonis2() if the user adds the `by="margin"` parameter. Option 3: Just use adonis2() with `by="terms"` as the default which is identical to current adonis(), as shown below, and expose the `by` parameters so users can switch  when needed. **My vote is Option 3 by far**.


## with adonis()

```r

library(tidyverse)
library(vegan)
library(broom)
data(dune)
data(dune.env)

set.seed(2022) 
adonis(dune ~ Management+A1, data = dune.env) %>% #build model
.$aov.tab %>% #grab summary table
broom::tidy() #broom's tidy cleans up results in a tibble. The warnings can be ignored.

'adonis' will be deprecated: use 'adonis2' instead
# A tibble: 4 × 7
  term          df SumsOfSqs MeanSqs F.Model    R2 p.value
  <chr>      <dbl>     <dbl>   <dbl>   <dbl> <dbl>   <dbl>
1 Management     3     1.47    0.490    3.07 0.342   0.004
2 A1             1     0.441   0.441    2.77 0.103   0.023
3 Residuals     15     2.39    0.159   NA    0.556  NA    
4 Total         19     4.30   NA       NA    1      NA    
Warning message:
In tidy.anova(.) :
  The following column names in ANOVA output were not recognized or transformed: SumsOfSqs, MeanSqs, F.Model, R2
```

## with adonis2, by="terms"

```r
set.seed(2022); adonis2(dune ~ Management+A1, data = dune.env, by="terms") %>% 
broom::tidy(.)

# A tibble: 4 × 6
  term          df SumOfSqs    R2 statistic p.value
  <chr>      <dbl>    <dbl> <dbl>     <dbl>   <dbl>
1 Management     3    1.47  0.342      3.07   0.004
2 A1             1    0.441 0.103      2.77   0.023
3 Residual      15    2.39  0.556     NA     NA    
4 Total         19    4.30  1         NA     NA    
Warning message:
In tidy.anova(.) :
  The following column names in ANOVA output were not recognized or transformed: SumOfSqs, R2
```

## And finally with adonis2, by="margin"

```r
set.seed(2022)
adonis2(dune ~ Management+A1, data = dune.env, by="margin") %>% 
broom::tidy(.)

# A tibble: 4 × 6
  term          df SumOfSqs    R2 statistic p.value
  <chr>      <dbl>    <dbl> <dbl>     <dbl>   <dbl>
1 Management     3    1.19  0.276      2.48   0.005
2 A1             1    0.441 0.103      2.77   0.023
3 Residual      15    2.39  0.556     NA     NA    
4 Total         19    4.30  1         NA     NA    
Warning message:
In tidy.anova(.) :
  The following column names in ANOVA output were not recognized or transformed: SumOfSqs, R2
```

A couple of notes here. The warnings here are related to the `tidy` function I used. I'm piping the adonis output into broom::tidy() as this gives nice tibbles that are convenient to work with downstream. The `broom::tidy` is a very useful way to get various stats objects into tables. It isn't usually used with adonis outputs, thus the warnings, though they are not important imo, but I find the output very useful still and it may actually help with Justine's hope of getting this into an artifact format. Adonis2() also doesn't require fetching the `aov.tab` object so one less thing to worry about.

@ebolyen I have some minor suggestions about those awesome plots over on the q2-fmt page you linked. Should I comment there?

**Comment by mestaki** [comment]
Looks like I was wrong about adonis2() being introduced to vegan only after 2.5_7 and after, as Jari Oksanen mentioned in the above Q2 forum post, this has been around since 2.4-0, so updating this will be even simpler, just a matter of updating the existing syntax to `adonis2` and exposing 1 or 2 additional parameters, which we can discuss here.

**Comment by colinbrislawn** [comment]
`broom::tidy()` does not support any vegan objects, [but ggvegan does!](https://github.com/gavinsimpson/ggvegan)

Thank you all for working on this problem. I'm excited for [q2-fmt](https://github.com/qiime2/q2-fmt)

**Comment by gregcaporaso** [comment]
Other adonis related feature requests that should be addressed at the same time as this one:
#303  
#242 
#243

**Comment by nbokulich** [comment]
`adonis` is now deprecated it seems, so there is some urgency to upgrade to `adonis2`

the other issues linked here do not need to be solved at the same time. Rome was not built in a day 😅

**Comment by lizgehret** [comment]
Thanks @nbokulich, I'll bring this up in this week's eng meeting - this would be something we'd need to prioritize in the first couple of months of our current release cycle, since we are now halting any recipe updates 3 mos pre-upcoming release, to make community developers' lives easier.

cc @gregcaporaso @ebolyen

**Comment by nbokulich** [comment]
Hey Liz, that makes a lot of sense! I don't think this requires a recipe
change — afaik `adonis2` is just another function in the `vegan` package.

**Comment by colinbrislawn** [comment]
[`adonis2` was added in 2.4 and depreciated in 2.6](https://github.com/vegandevs/vegan/releases?q=adonis2&expanded=true)

https://github.com/qiime2/q2-diversity/blob/28c61bebba89c384f671f7a589dfba3c759d6259/ci/recipe/meta.yaml#L52

We should be good to go!

I don't hate the idea of bumping vegan, but it's stable, old-school software this may be unneeded!

**Comment by lizgehret** [comment]
> Hey Liz, that makes a lot of sense! I don't think this requires a recipe change — afaik `adonis2` is just another function in the `vegan` package.

Yeah we don't directly depend on adonis, but if our conda env resolves vegan to >=2.6 (which could be possible in the future since we just have a min pin on vegan), we'll just have to make sure that we make any changes to be compatible with adonis2. So we may want to get ahead of that and bump that min pin, and then just make any needed changes within diversity to call adonis2 (instead of adonis).

**Comment by ARW-UBT** [comment]
Hello, 
I have seen this line
res <- adonis2(formula, by='terms', data=sample.md, permutations=perms, parallel=njobs)
at https://github.com/qiime2/q2-diversity/blob/dev/q2_diversity/_beta/adonis_assets/run_adonis.R#L35

Does this mean that adonis2 is used already in q2-amplicon-2025.4 and later by `qiime diversity adonis `?
Just want to make sure to add the correct version to a manuscript.
Best regards

**Comment by lizgehret** [comment]
Hi @ARW-UBT,

Thanks for reaching out. As of 2026.4 (our upcoming release, scheduled for 15 April), all calls to adonis will be replaced with adonis2. For versions 2026.1 and earlier, we were still on adonis. I'm going to go ahead and close this issue as completed.

Cheers 🦎
